### DISCUSSION TITLE
connect gui to background program

### DISCUSSION THREAD

**Discussion by mrwonderfulness** [pyqtgraph/pyqtgraph#1745] [Q&A]
Hello, pyqtgraph noob here...
i have a python program running non-stop and it does not have a gui. i would like to sometimes "connect" to it and view the data it is producing, and perhaps even interact with it to stop it or change some parameters. so, my Qt/pyqgraph will only be used occasionally,  but the background process cannot stop. it is constantly fetching data from a websocket and to close it will mean to lose data.
how, in general, should i approach this? i would like to keep gui code out of the main program, though i realize there must be some kind of signal that will need to be sent in order to communicate with the gui (if the gui is running).
i guess what confuses me is the gui application loop...the program i am running in the background i initially tried to add pyqtgrapgh code to, but i have a while loop that is constantly waiting for/fetching new data, and this needs to happen before the app.exec_(), but then the problem is that the control flow never gets to that last line...
there must be a standard way of dealing with this. thanks.

**Comment by j9ac9k** [comment]
Hi @mrwonderfulness 

I think having the pyqtgraph data vis portion be separate from the non-stop running application that collects data over web-sockets is a good plan.  I'm not versed in good methods of transmitting data between two applications; so I can't give much guidance in that regard.

Depending on the amount of data we're talking, the frequency we're updating, and how "live" the data needs to be, I would likely consider having the application doing the data collection write data out to a sqlite database, and have the GUI/plotting application read that.  You can setup a [QFileSystemWatcher](http://doc.qt.io/qt-5/qfilesystemwatcher.html) instance to watch for the file that gets updated, and on each update, you replot/redraw.

Again, this is not an area I have much knowledge in, so I'm curious if you get other suggestions, what they would be.

**Reply by mrwonderfulness** [reply]
thanks for your reply, but even in this scenario, how would i get the loop that checks the database to run within the qt application loop (not using a qt timer). is this where threading comes in?

**Reply by j9ac9k** [reply]
The QFileSystemWatcher has a fileChanged signal http://doc.qt.io/qt-5/qfilesystemwatcher.html#fileChanged that emits a signal when a file is changed, I would aim it at the SQLite database, and whenever that file changes, have the GUI app re-read the contents and update them (or a partial read depending on the size of the data).  If you do this, I highly recommend using pandas to read the sqlite database as that's just about a 1-liner, and then feed the numpy arrays within each pandas columns to pyqtgraph's plotting methods.

**Reply by j9ac9k** [reply]
An alternative to sqlite would be h5py (which pyqtgraph actually has native support for).

On the application doing the gathering, I would keep writing/modifying a HDF5 file, and have the GUI application read/update, again, likely using QFileSystemWatcher

**Reply by mrwonderfulness** [reply]
Could you suggest a good way to run a loop inside the Qt application loop? I notice there are many examples that use Qt timer, but i would like for new incoming data to be happing INSIDE the Qt app loop. so, basically...
1. a loop is running that is listening for new data
2. when data arrives, it gets processed
3. the data then updates the plot

i think this relates to "worker" threads and Qthread...but can't seem to find a good example of it implemented with pyqtgraph.

is there an example with pyqtgraph that you could direct me to?

thanks.

**Reply by pijyoi** [reply]
Following is a program that plots incoming data from a udp socket.
As a demonstration, data is produced by reading samples from the sound card.
For convenience of the example, both the udp sender and udp receiver are run within the same process.

No threading is used as the various objects all run off the Qt event loop.

```python
from PySide2.QtMultimedia import QAudio, QAudioDeviceInfo, QAudioFormat, QAudioInput
from PySide2 import QtNetwork
import pyqtgraph as pg
import numpy as np

app = pg.mkQApp()
plt = pg.PlotWidget()
plt.show()

Fsamp = 8000
plt.setYRange(-4000, 4000)
plotdata = np.ones(Fsamp*10, dtype=np.int16)
curve = plt.plot(plotdata)

def update_plot(data):
    samples = np.frombuffer(data, dtype=np.int16)
    plotdata[:-samples.size] = plotdata[samples.size:]
    plotdata[-samples.size:] = samples
    curve.setData(plotdata)

def sock_readable():
    while rsock.hasPendingDatagrams():
        datagram = rsock.receiveDatagram()
        update_plot(datagram.data())

rsock = QtNetwork.QUdpSocket()
rsock.bind(QtNetwork.QHostAddress("127.0.0.1"), 12345)
rsock.readyRead.connect(sock_readable)

ssock = QtNetwork.QUdpSocket()
ssock.connectToHost("127.0.0.1", 12345)

def audio_readable():
    data = qiodev.readAll()
    if not data:
        return
    # update_plot(data)
    ssock.write(data)

fmt = QAudioFormat()
fmt.setChannelCount(1)
fmt.setSampleRate(Fsamp)
fmt.setSampleSize(16)
fmt.setCodec("audio/pcm")
fmt.setByteOrder(QAudioFormat.LittleEndian)
fmt.setSampleType(QAudioFormat.SignedInt)

ainfo = QAudioDeviceInfo.defaultInputDevice()
print(ainfo.deviceName())

ainput = QAudioInput(fmt, None)
qiodev = ainput.start()
qiodev.readyRead.connect(audio_readable)

app.exec_()
```

**Reply by mrwonderfulness** [reply]
how does the Qt loop know how quickly to update? is it because the audio/network libraries are all Qt based? What if you had data being generated at varying speeds from a source that is not controlled by Qt. For example, let's say you have data from a sensor or market data or similar that  needs it's own loop to check for incoming data...that had to be listening in a while True: loop?

**Reply by pijyoi** [reply]
The Qt event loop is event/data driven. In the example, the audio source is the sensor and it emits a signal when it has data to be drained.

You could remove the audio and send socket from the example. Then in a separate non-Qt process that is running your while True: loop, you could send udp packets to be plotted. 

The IPC doesn't have to be using udp. I use it here for the simplicity.

**Reply by mrwonderfulness** [reply]
thank you very much for your reply, but i'm finding it hard to understand from your code how you are using the incoming data as the event that updates the graph.

I have data coming in sporatically, sometimes nothing per second, sometimes several hundred....and i need for THOSE events to be updating my graph, not a timer, and not in "chunks" per a timer interval....i.e. i want them to be plotted as they arrive.

I think part of my difficulty is understanding how the networking code is working in your example.

**Reply by pijyoi** [reply]
```python
rsock.readyRead.connect(sock_readable)
```
When udp packets arrive _asynchronously_, the readyRead signal is emitted and the ```sock_readable``` function will be called. https://doc.qt.io/qt-5/qiodevice.html#readyRead

```sock_readable``` drains the udp socket and calls ```update_plot``` to plot the new data.

In an event-driven framework, you need to write your code such that the event loop does not get blocked by long-running callbacks (or slots in Qt parlance).

**Reply by mrwonderfulness** [reply]
wouldn't it be better to have the work done in a separate thread or process? e.g. by using Qthread?

**Reply by mrwonderfulness** [reply]
ok, i see, readReady has built-in async operation like Qprocess. i have to build it from scratch so this won't work in my case. i want to find out the best way to prevent the main qt loop to not interfere with the worker,  which runs continously and needs to update my pyqtgraph chart several times each second

**Reply by j9ac9k** [reply]
This is somewhat difficult for us to comment on as you understand your system better than we do.  I think pijyoi and I were of the mindset that whatever application is doing the collection would need to have a slight modification to emit the new data via udp packets; and the GUI application, a completely different application would receive those packets and handle the display.


I haven't seen anything here to suggest why this wouldn't work; but as I said, you know your system/requirements better then we do.

Regarding the use of multiple threads, that's certainly doable but I've found that going multithreaded is only helpful when you are able to be very very specific with the problem you're trying to solve.

**Reply by mrwonderfulness** [reply]
It probably would work, but I am not sure to implement it from scratch the way pijyoi handles it by default (asyncronously). In the docs of readReady it states that that async behaviour is built in, but what if you were using a custom tool from scratch that you wanted to run outside the Qt loop? 

I already have a program that reads from a socket and saves to a redis database. in a second program i have this reading data published from redis, which is then processed. this data is coming several times per second and i want to view it in pyqtgraph, having THE EVENT of it arriving updating the graph, not a Qt timer. this process is constantly running in the background and i need for it to be able to communicate with pyqtgraph. 

i realize there is the QApplication.processEvents() option, but if you read here, it is the "dumb approach":
https://www.mfitzp.com/tutorials/multithreading-pyqt-applications-qthreadpool/#the-dumb-approach

**Reply by pijyoi** [reply]
Qt-based plotting process with audio and send sock removed.
```python
from PySide2 import QtNetwork
import pyqtgraph as pg
import numpy as np

app = pg.mkQApp()
plt = pg.PlotWidget()
plt.show()

plt.setYRange(-4000, 4000)
plotdata = np.ones(8000, dtype=np.int16)
curve = plt.plot(plotdata)

def update_plot(data):
    samples = np.frombuffer(data, dtype=np.int16)
    plotdata[:-samples.size] = plotdata[samples.size:]
    plotdata[-samples.size:] = samples
    curve.setData(plotdata)

def sock_readable():
    while rsock.hasPendingDatagrams():
        datagram = rsock.receiveDatagram()
        update_plot(datagram.data())

rsock = QtNetwork.QUdpSocket()
rsock.bind(QtNetwork.QHostAddress("127.0.0.1"), 12345)
rsock.readyRead.connect(sock_readable)

app.exec_()
```

Non-Qt process running in a forever loop
```python
import time
import socket
import random
import numpy as np

ssock = socket.socket(type=socket.SOCK_DGRAM)
ssock.connect(("127.0.0.1", 12345))

while True:
    # simulate sporadic data
    msec = random.randint(0, 100)
    time.sleep(msec * 0.001)
    noise = np.random.normal(scale=100, size=(320,))
    data = noise.astype(np.int16)
    ssock.send(data)
```

**Reply by ntjess** [reply]
From the problem description, pijyoi's solution seems like the way to go. However, if you _really_ didn't want to execute the gui loop, you can get a poor man's version by calling `processEvents` repeatedly instead. Then, your python loop can still execute. Again, this is not a robust or recommended solution, but it is an alternative.

```python
import pyqtgraph as pg
from pyqtgraph.Qt import QtWidgets, QtGui, QtCore

app = pg.mkQApp()
plt = pg.PlotWidget()
spi = pg.ScatterPlotItem()
plt.addItem(spi)
plt.show()

def mkUpdate(*data):
  spi.setData(*data)
  plt.update()
  

QtCore.QTimer.singleShot(3000, lambda: mkUpdate([1,2,3], [7,8,9]))
QtCore.QTimer.singleShot(5000, lambda: mkUpdate([2, 5, 6, 2], [6, 7, 4, 1]))

while True:
  # get your data here, etc.
  # This will update your plot, but if it's not called often enough your window will hang
  # I imagine you may want to call this at the end of your loop that gets socket data?
  app.processEvents()
```

**Comment by rhuygen** [comment]
Hi @mrwonderfulness 

I have a similar setup where I have a server running that reads out housekeeping and data from a hardware device and saves the data into HDF5 and CSV files. Then I have a number of GUI applications that visualize the data and the housekeeping. 

I have several 'streams' that interact with the server app. One stream for commanding the server, e.g. change device settings, change update frequency, quit the server app, etc. Another stream to monitor the server app in terms of memory and CPU usage, etc., and another stream for distributing data that I want to visualize in real-time. 

I use ZeroMQ as a messaging protocol and distribute the data. The commanding stream is a REQ-REP protocol, the monitoring is a PUB-SUB protocol, and the data distribution is a PUSH-PULL protocol.

For the `pyqtgraph` visualization app, I have a separate thread that reads the data from the data distribution stream of the server and sends specific signals to the main GUI app to update and display the data.

The data stream to visualize peaks to 20MB/s and I have no performance problems so far.

Hope this helps.

Here is a screenshot of the (unfinished) GUI:
![image](https://user-images.githubusercontent.com/6662864/123765089-aeef3000-d8c5-11eb-9559-6f66e3379426.png)

**Reply by mrwonderfulness** [reply]
Very sorry to have missed your response and not replied. Thank you for your time. It's interesting that I also ended up using ZeroMQ for this project. It allowed me to make things more modular just as you describe above. Thanks again and apologies for my 1+ year late reply! I'm a self-taught programmer and sometimes it feels overwhelming but by doing tests, digging around the internet and asking questions...somehow it all comes together. Best regards.
